src/plugins/intel_cpu/src/nodes/multiclass_nms.cpp

```diff
@@ -411,182 +387,210 @@ float MultiClassNms::intersectionOverUnion(const float* boxesI, const float* box
-
-                    bool box_is_selected = true;
-                    for (int idx = static_cast<int>(fb.size()) - 1; idx >= currBox.suppress_begin_index; idx--) {
-                        float iou = intersectionOverUnion(&boxesPtr[currBox.idx * 4], &boxesPtr[fb[idx].box_index * 4], m_normalized);
-                        currBox.score *= func(iou, adaptive_threshold);
-                        if (iou >= adaptive_threshold) {
-                            box_is_selected = false;
-                            break;
-                        }
-                        if (currBox.score <= m_scoreThreshold)
-                            break;
-                    }
-
-                    currBox.suppress_begin_index = fb.size();
-                    if (box_is_selected) {
-                        if (m_nmsEta < 1 && adaptive_threshold > 0.5) {
-                            adaptive_threshold *= m_nmsEta;
-                        }
-                        if (currBox.score == origScore) {
-                            fb.push_back({currBox.score, batch_idx, class_idx, currBox.idx});
-                            continue;
-                        }
-                        if (currBox.score > m_scoreThreshold) {
-                            sorted_boxes.push(currBox);
-                        }
-                    }
+
+        const int num_boxes_per_class = (attr_nms_top_k_ < 0) ? num_boxes_selected : std::min(attr_nms_top_k_, num_boxes_selected);
+        if (num_boxes_per_class == 0) {
+            workbuffers_->num_boxes_per_batch_and_class[batch_idx][class_idx] = 0;
+            return;
+        } else if (num_boxes_selected / num_boxes_per_class < 10) {
+            parallel_sort(boxes, boxes + num_boxes_selected, [](const Box& l, const Box& r) {
+                return (l.score > r.score || ((l.score == r.score) && (l.box_idx < r.box_idx)));
+            });
+        } else {
+            std::partial_sort(boxes, boxes + num_boxes_per_class, boxes + num_boxes_selected,
+                [](const Box& l, const Box& r) {
+                    return (l.score > r.score || ((l.score == r.score) && (l.box_idx < r.box_idx)));
+                });
+        }
+
+        num_boxes_selected = 0;
+        float iou_threshold = attr_iou_threshold_;
+        for (size_t i = 0; i < num_boxes_per_class; i++) {
+            bool box_is_selected = true;
+            for (int j = num_boxes_selected - 1; j >= 0; j--) {
+                const float iou = intersection_over_union(&input_boxes[boxes[i].box_idx * 4],
+                                                          &input_boxes[boxes[j].box_idx * 4],
+                                                          attr_normalized_);
+                if (iou >= iou_threshold) {
+                    box_is_selected = false;
+                    break;
                 }
+                if (boxes[i].score == attr_score_threshold_)    // TODO: bug in reference impl?
+                    break;
```

Comment:
See https://github.com/openvinotoolkit/openvino/blob/master/src/core/reference/src/runtime/reference/multiclass_nms.cpp#L296

Reply:
why do you think it is a bug in reference impl?

Reply:
I don't think it's a bug, I rather suspect it's a bug.

We can safely replace operator '<=' with '==' in line [L296](https://github.com/openvinotoolkit/openvino/blob/master/src/core/reference/src/runtime/reference/multiclass_nms.cpp#L296), because the condition `next_candidate.score < attrs.score_threshold` is always `false`. The only place where we change `next_candidate.score`, so that this condition could become `true`, is when we reset score to zero in line [L289](https://github.com/openvinotoolkit/openvino/blob/master/src/core/reference/src/runtime/reference/multiclass_nms.cpp#L289). But when this happens, the condition in line [L291](https://github.com/openvinotoolkit/openvino/blob/master/src/core/reference/src/runtime/reference/multiclass_nms.cpp#L291) is evaluated to `true`, `should_hard_suppress` becomes `true` and the `for` loop is interrupted in line [L293](https://github.com/openvinotoolkit/openvino/blob/master/src/core/reference/src/runtime/reference/multiclass_nms.cpp#L293).

**So, I suspect that in line 296 the original intention was to use '<' operator instead of '<='.** Because why otherwise  would someone want to check `box intersection area` only with the last selected box when score is equal to `score threshold`, and not with all previously selected boxes like we do when score is strictly greater then `score threshold`? It's suspicious.

It's not the only `dead` code I see in reference implementation. Line 279, 307, lines 311-313 can be safely removed. The condition in line 307 is always evaluated to `true` and execution  flow never reaches lines 311-313.

Reply:
Sounds reasonable. Can you run paddlepaddle frontend unit test (with gtest_filter=*multiclassnms*) locally to check if you doubt there might be a bug in algorithm implementation?

Reply:
But the current reference impl won't depends on L289, i.e. no matter what "func" is.

Reply:
I see the same about L289. It either does nothing or resets `next_candidate.score` to zero if `iou > adaptive_threshold`. But when `iou > adaptive_threshold`, the `next_candidate` box (with now zero score) is just filtered out (not used anymore farther in algorithm).

Reply:
@ceciliapeng2011 
paddlepaddle frontend unit tests compare expected outputs from numpy binary files with CPU plugin outputs.
These tests succeed with or without fixing this bug in CPU plugin. 
On the other hand, some benchmark tests I've implemented for this PR are failing with the fix applied in CPU plugin because reference implementation gives different outputs.
The fix I'm talking about is to remove these to lines:
```
437                if (boxes[i].score == attr_score_threshold_)    // TODO: bug in reference impl?
438                      break;
```
Actually, I've added this 2 lines myself to align implementation with reference and fix the issue with my benchmark tests. This was done in second commit to this PR.
My conclusion is that we can fix reference without breaking paddlepaddle compatibility. It may or may not affect other plugins (at least CPU functional tests succeed with or without fixing this bug in CPU plugin).

Reply:
@ceciliapeng2011 
Urelated FYI: The following test from paddlepaddle frontend tests is failing for me locally: `PaddleJsonConfigTest/FrontEndJsonConfigTest.testAddJsonConfigExtension/paddle_relu_relu_pdmodel`